Allow NULLs in CASE and transform

aiven-sal opened the pull request:
`CASE` uses `caseWithExpression` behind the scenes, which in turn calls `transform`.
To be able to use `CASE` without an `ELSE` branch we need `transform` to be able to deal with NULL default values.

To achieve this I completely rewrote the transform function. Some old tests are broken by these changes, but it looks like those tests were wrong in the first place: what they are testing doesn't seem to be coherent with the documented behaviour of `transform`.

Closes #7237

<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/run_check.py
-->
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix a bug that prevented the use of `CASE` without an `ELSE` branch and extended `transform` to deal with more types. Also fix some bugs that made transform() return incorrect results when decimal types were mixed with other numeric types.

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

aiven-sal committed:
Functions: allow transform to deal with NULLs and simplify the code

The function has been completely rewritten to be able to handle
parameters in a generic way an support NULLs.

This is Part of the fix for #7237

aiven-sal committed:
Functions: allow caseWithExpression to correctly deal with NULLs

Closes #7237

aiven-sal committed:
tests: fix incorrect tests

Some tests were checking for some undocumented and counter-intuitive
behaviour in function trasform.
The new function doesn't behave in that way and those tests fail.

aiven-sal committed:
tests: add new tests for transform and case

antonio2368 commented on tests/queries/0_stateless/02497_if_transform_strings_to_enum.sql:
```diff
@@ -33,13 +33,13 @@ SELECT transform(number, [2, 4, 6], ['google', 'censor.net', 'yahoo'], 'other')
-SELECT transform(number, NULL, ['google', 'censor.net', 'yahoo'], 'other') FROM system.numbers LIMIT 10;
-EXPLAIN SYNTAX SELECT transform(number, NULL, ['google', 'censor.net', 'yahoo'], 'other') FROM system.numbers LIMIT 10;
-EXPLAIN QUERY TREE run_passes = 1 SELECT transform(number, NULL, ['google', 'censor.net', 'yahoo'], 'other') FROM system.numbers LIMIT 10;
+SELECT transform(number, NULL, ['google', 'censor.net', 'yahoo'], 'other') FROM system.numbers LIMIT 10; -- { serverError 43 }
```
This seems wrong, you are throwing exception when it failed to apply some optimization to a single transform.
What if you have in the same query `transform` for which this optimization can be applied?
Better fix IMO would be to modify `ConvertStringsToEnumVisitor` and `IfTransformStringsToEnumPass`.

antonio2368 replied:
ah yes that also makes sense 👍

CurtizJ commented:
Unfortunately you cannot remove all specializations and leave only generic version. The performance will degrade significantly. You have to add a generic version in addition.

Manually tested query from performance tests:
```
clickhouse benchmark <<< "SELECT transform(number, [2, 4, 6], ['google', 'yandex', 'yahoo'], 'other') as x FROM numbers(10000000) GROUP BY x FORMAT Null" --cumulative
```

Your PR:
```
localhost:9000, queries 63, QPS: 6.843, RPS: 68478768.709, MiB/s: 522.452, result RPS: 0.000, result MiB/s: 0.000.

0.000%          0.143 sec.
10.000%         0.143 sec.
20.000%         0.144 sec.
30.000%         0.144 sec.
40.000%         0.145 sec.
50.000%         0.145 sec.
60.000%         0.146 sec.
70.000%         0.147 sec.
80.000%         0.148 sec.
90.000%         0.150 sec.
95.000%         0.152 sec.
99.000%         0.153 sec.
99.900%         0.155 sec.
99.990%         0.155 sec.
```

Current master:
```
localhost:9000, queries 91, QPS: 9.530, RPS: 95372740.442, MiB/s: 727.636, result RPS: 0.000, result MiB/s: 0.000.

0.000%          0.100 sec.
10.000%         0.101 sec.
20.000%         0.102 sec.
30.000%         0.102 sec.
40.000%         0.102 sec.
50.000%         0.103 sec.
60.000%         0.103 sec.
70.000%         0.103 sec.
80.000%         0.104 sec.
90.000%         0.105 sec.
95.000%         0.112 sec.
99.000%         0.155 sec.
99.900%         0.158 sec.
99.990%         0.158 sec.
```

The difference is about 40%.

aiven-sal committed:
Improve performance

aiven-sal committed:
Fix test

aiven-sal committed:
Fix old test

the test expected caseWithExpression to return NULL every time
there was a NULL as an argument, regardless of the other arguments'
values.

The original intent of this test is not clear to me.

CurtizJ commented:
> Do you remember exactly what commits did you test?

No, it was the latest master. Anyway let's wait the result of perf tests in CI.

aiven-sal committed:
Re-add specializations to improve performance

aiven-sal committed:
Rename tests

aiven-sal committed:
Add new regression tests

aiven-sal committed:
Add new test about mixing decimals and ints

This test is broken with the old code

aiven-sal committed:
Restart CI

aiven-sal committed:
Fix warning

aiven-sal committed:
Add missing check for argument type

aiven-sal committed:
Fix output ordering in test

aiven-sal committed:
Add a test for another bug fixed in this PR

aiven-sal committed:
Rename variables with misleading names

aiven-sal committed:
Fix more issues with Decimals

These issues were already present in the original code

aiven-sal committed:
Prevent types too large for UInt64 from being used

aiven-sal committed:
Restart CI

CurtizJ reviewed (approved):
LGTM

CurtizJ commented:
Please, resolve the conflicts and I'll merge the PR then.

alexey-milovidov commented:
@CurtizJ Unfortunately, the Fuzzer has found a bug in this pull request:
```
milovidov-desktop :) SELECT caseWithExpr(arrayReduce(NULL, []), []);

SELECT caseWithExpr(arrayReduce(NULL, []), [])

Query id: 24445c4f-bb82-49dc-b025-8bc5c84fde6b


0 rows in set. Elapsed: 0.001 sec. 

Received exception:
Code: 49. DB::Exception: Value of type Array(Nothing) in memory is not of fixed size.: While processing caseWithExpr(arrayReduce(NULL, []), []). (LOGICAL_ERROR)
```

alexey-milovidov commented:
This PR has bad code style: cryptic variable names.

`dt` -> data_type
`dtsize` -> data_type_size

alexey-milovidov commented:
I will help...

alexey-milovidov commented:
#49616

alexey-milovidov commented:
#49625
